### PR TITLE
Handle replacements properly for runDev()

### LINKED ISSUE #68: xcaddy development mode fails if 'go.mod' includes 'replace' directive
### How to reproduce this issue:

Clone https://github.com/simnalamburt/xcaddybug and try to build with xcaddy.

```go
// main.go
package xcaddybug
import "example.com/stringutil"
import "github.com/caddyserver/caddy/v2"

func init() {
	caddy.RegisterModule(Example{})
}
type Example struct {}
func (Example) CaddyModule() caddy.ModuleInfo {
	return caddy.ModuleInfo{
		ID:  caddy.ModuleID(stringutil.Reverse("elpmaxe.oof")),
		New: func() caddy.Module { return new(Example) },
	}
}
```
```go
// go.mod
module github.com/simnalamburt/xcaddybug
go 1.17
replace example.com v0.0.1 => golang.org/x/example v0.0.0-20210811190340-787a929d5a0d
require (
	example.com v0.0.1
	github.com/caddyserver/caddy/v2 v2.4.3
)
```

### Error message:
```
2021/08/29 22:55:01 [INFO] Resolved relative replacement example.com=>golang.org/x/example v0.0.0-20210811190340-787a929d5a0d to /home/simnalamburt/workspace/xcaddybug/golang.org/x/example v0.0.0-20210811190340-787a929d5a0d
2021/08/29 22:55:01 [INFO] Temporary folder: /tmp/buildenv_2021-08-29-2255.2607006221
2021/08/29 22:55:01 [INFO] Writing main module: /tmp/buildenv_2021-08-29-2255.2607006221/main.go
2021/08/29 22:55:01 [INFO] Initializing Go module
2021/08/29 22:55:01 [INFO] exec (timeout=10s): /usr/sbin/go mod init caddy
go: creating new go.mod: module caddy
go: to add module requirements and sums:
        go mod tidy
2021/08/29 22:55:01 [INFO] Replace github.com/simnalamburt/xcaddybug => /home/simnalamburt/workspace/xcaddybug
2021/08/29 22:55:01 [INFO] exec (timeout=10s): /usr/sbin/go mod edit -replace github.com/simnalamburt/xcaddybug=/home/simnalamburt/workspace/xcaddybug
2021/08/29 22:55:01 [INFO] Replace example.com => /home/simnalamburt/workspace/xcaddybug/golang.org/x/example v0.0.0-20210811190340-787a929d5a0d
2021/08/29 22:55:01 [INFO] exec (timeout=10s): /usr/sbin/go mod edit -replace example.com=/home/simnalamburt/workspace/xcaddybug/golang.org/x/example@v0.0.0-20210811190340-787a929d5a0d
2021/08/29 22:55:01 [INFO] Pinning versions
2021/08/29 22:55:01 [INFO] exec (timeout=0s): /usr/sbin/go get -d -v github.com/caddyserver/caddy/v2
go: errors parsing go.mod:
/tmp/buildenv_2021-08-29-2255.2607006221/go.mod:7: replacement module directory path "/home/simnalamburt/workspace/xcaddybug/golang.org/x/example" cannot have version
2021/08/29 22:55:01 [ERROR] exit status 1
```

### References
- https://github.com/caddyserver/xcaddy/blob/b8a0d4664/cmd/xcaddy/main.go#L186-L209

### PR DESCRIPTION
**Pull request by simnalamburt**
Closes #68 

1.  Handle all possible replacements properly for `runDev()`. Replacement targets are not always paths.
2.  Parse replacement info from `go list -m -json all`. It's more robust in this way.

I have tested all possible cases with replacements like below. So hopefully replacements won't cause any issue further at least for the `runDev()`.

```go
replace (
	replacetest1 v1.2.3 => golang.org/x/example v0.0.0-20210811190340-787a929d5a0d
	replacetest2 => golang.org/x/example v0.0.0-20210407023211-09c3a5e06b5d
	replacetest3 v1.2.3 => ./fork1
	replacetest4 => ./fork2
	replacetest5 v1.2.3 => /home/username/fork3
	replacetest6 => /home/username/fork4
)
```

Use forked version of xcaddy if you want to test this PR:
```bash
go install github.com/simnalamburt/xcaddy/cmd/xcaddy@466a1a4b86
```

#### References
- https://pkg.go.dev/cmd/go/internal/list#pkg-variables

### CODE REVIEW

**Comment by Mohammed90** [comment]
Thank you, @simnalamburt! If you don't mind, since you've stumbled upon an edge case, do you mind adding few tests to further validate the logic? We'd hate to break existing users while fixing a bug.

**Comment by simnalamburt** [comment]
I’m not familiar with xcaddy internals (this is my first day with xcaddy,
and I’ve never used go for last 3 years) so actually I don’t know how to so
it. Can you give me some brief guide of how to test this?

2021년 8월 30일 (월) 04:46, Mohammed Al Sahaf ***@***.***>님이 작성:

**Comment by mholt** [comment]
I think unit tests would go a long way, at least. Might have to split some of the logic into a separate function to do that effectively. Basically checking that the input and outputs are as expected. https://pkg.go.dev/testing is your friend. :+1:

**Comment by simnalamburt** [comment]
Maybe I can separate executing `go list -m -json all` part and parsing the output part, and then unit test the parsing part. Thanks @mholt! :)

**Comment by simnalamburt** [comment]
I think I’ve done some basic testing :) Please feel free to take a look!

**Comment by Mohammed90** [comment]
> Omg the test was UNIX specific 🤦 hang on

Yay for tests 🙂

**Comment by Mohammed90** [comment]
> Omg the test was UNIX specific 🤦 hang on

Let me try to make the issue clearer. Note the slashes here:

<table>
<tr>
<th>
Expected
</th>
<th>
Got
</th>
</tr>

<tr>

<td>
<pre>
[
	{
		replacetest1@v1.2.3
		golang.org/x/example@v0.0.0-20210811190340-787a929d5a0d
	}
	{
		replacetest2@v0.0.1 
		golang.org/x/example@v0.0.0-20210407023211-09c3a5e06b5d
	} 
	{
		replacetest3@v1.2.3 
		\home\work\module\fork1
	} 
	{
		github.com/simnalamburt/module 
		\home\work\module
	} 
	{
		replacetest4@v0.0.1 
		\srv\fork2
	} 
	{
		replacetest5@v1.2.3 
		\home\work\module\fork3
	}
]
</pre>
</td>

<td>
<pre>
[
	{
		replacetest1@v1.2.3 
		golang.org/x/example@v0.0.0-20210811190340-787a929d5a0d
	}
	{
		replacetest2@v0.0.1 
		golang.org/x/example@v0.0.0-20210407023211-09c3a5e06b5d
	}
	{
		replacetest3@v1.2.3 
		\home\work\module\fork1
	}
	{
		github.com/simnalamburt/module 
		/home/work/module
	}
	{
		replacetest4@v0.0.1 
		\home\work\module\srv\fork2
	}
	{
		replacetest5@v1.2.3 
		\home\work\module\fork3
	}
]
</pre>
</td>

</tr>
</table>

**Comment by simnalamburt** [comment]
OK I think I fixed windows build, but now I failed Go 1.16 build due to the use of `//go:build` style build constraints.

**Comment by Mohammed90** [comment]
> OK I think I fixed windows build, but now I failed Go 1.16 build due to the use of `//go:build` style build constraints.

For this you have to use the older style `// +build windows` just below the newer style. Have them both in the same file just below each other.

**Comment by simnalamburt** [comment]
Note: I tried not to split main_unix_test.go and main_windows_test.go, but their JSON looks too different so it was really hard to generate two JSON in one test code.

**Comment by francislavoie** [comment]
If only Go supported the ternary condition operator :joy:

File: cmd/main.go [hunk 1]
```diff
@@ -277,6 +251,71 @@ func runDev(ctx context.Context, args []string) error {
 	return cmd.Wait()
 }
 
+func parseGoListJson(out []byte) (currentModule, moduleDir string, replacements []xcaddy.Replace, err error) {
+	var unjoinedReplaces []int
+
+	decoder := json.NewDecoder(bytes.NewReader(out))
+	for {
+		var mod map[string]interface{}
```
Comment (Mohammed90) [inline cmd/main.go:259]: You can avoid the map-of-string-to-empty-interface with a struct. The Go team defines the shape of the struct and JSON: https://golang.org/ref/mod (search for `type Module struct`). This will save you from having to do type assertion and string based access to the fields.

**Review by Mohammed90 — CHANGES_REQUESTED** [review]
This looks good! One minor nitpick then it's all right.

File: cmd/main.go (hunk 1 above)
Comment (simnalamburt) [reply cmd/main.go]: Got it!

**Comment by simnalamburt** [comment]
I changed `map[string]interface{}` into `Module` and the CI has [passed in my fork](https://github.com/simnalamburt/xcaddy/actions/runs/1189494024).

**Review by Mohammed90 — APPROVED** [review]
Had to handle one last minor nitpick. Perfecto. Thanks @simnalamburt!
